File: ising2d.py
import random
import math
import time
import os

# Parameters
NX = 64
NY = 64
ntherm = 0
VisualDisplay = 1
SleepTime = 100000  # in microseconds

# Initialize the lattice with boundary spins
spin = [[0 for _ in range(NY + 2)] for _ in range(NX + 2)]

def update_spin(s, env):
    """Do a metropolis update on a spin s whose environment is env"""
    spin = s[0]
    newspin = 1 if random.random() < 0.5 else -1
    DeltaBetaE = -(newspin - spin) * env
    if DeltaBetaE <= 0 or random.random() < math.exp(-DeltaBetaE):
        s[0] = newspin

def sweep(beta, h):
    """Sweep through all lattice sites"""
    for nx in range(1, NX + 1):
        for ny in range(1, NY + 1):
            environment = (beta * (spin[nx][ny-1] + spin[nx][ny+1] + 
                                 spin[nx-1][ny] + spin[nx+1][ny]) + h)
            # Using list to simulate pointer
            spin_ref = [spin[nx][ny]]
            update_spin(spin_ref, environment)
            spin[nx][ny] = spin_ref[0]

def initialize_hot():
    """Initialize lattice with random spins"""
    print("Initializing system")
    for nx in range(NX + 2):
        for ny in range(NY + 2):
            if nx == 0 or nx == NX + 1 or ny == 0 or ny == NY + 1:
                spin[nx][ny] = 0
            else:
                spin[nx][ny] = 1 if random.random() < 0.5 else -1

def magnetization():
    """Calculate average magnetization"""
    nmag = sum(spin[nx][ny] for nx in range(1, NX + 1) 
              for ny in range(1, NY + 1))
    return nmag / (NX * NY)

def display_lattice(sweep_number):
    """Display the lattice configuration"""
    if SleepTime > 0:
        os.system('cls' if os.name == 'nt' else 'clear')
    
    for nx in range(1, NX + 1):
        for ny in range(1, NY + 1):
            print('X' if spin[nx][ny] == 1 else '-', end='')
        print()
    
    if sweep_number >= 0:
        print(f"sweep {sweep_number}:   magnetization <sigma> = {magnetization():.6f}")
    
    if SleepTime > 0:
        time.sleep(SleepTime / 1_000_000)  # Convert microseconds to seconds
    else:
        print()

def main():
    print(f"Program generates a thermal ensemble a 2D Ising model of "
          f"{NX}x{NY} spins with free boundary conditions.\n")
    
    random.seed(int(time.time()))
    
    nsweep = int(input("Enter total number of configurations generated:\n"))
    h = float(input("Enter value of magnetic field parameter h:\n"))
    beta = float(input("Enter temperature parameter beta (= 1/kT):\n"))
    
    initialize_hot()
    
    # Thermalization sweeps
    print(f"Thermalizing system, {ntherm} sweeps")
    for n in range(ntherm):
        sweep(beta, h)
        if VisualDisplay and n % (ntherm//10 if ntherm > 0 else 1) == 0:
            display_lattice(-1)
            print(f"Thermalization sweep {n}")
            time.sleep(1)
    
    # Main sweeps
    nmag = ntotal = 0
    for n in range(nsweep):
        if VisualDisplay:
            display_lattice(n)
        sweep(beta, h)
        for nx in range(1, NX + 1):
            for ny in range(1, NY + 1):
                nmag += spin[nx][ny]
                ntotal += 1
    
    if VisualDisplay:
        display_lattice(nsweep)
    print(f"Average Magnetization: <s> = {nmag/ntotal:.6f}")

if __name__ == "__main__":
    main()
    

File: ising2d.cpp
/* 
******************************************************************************
* ising2d.c                                                                  *
* =========                                                                  *
*    Uses Metropolis algorithm to generate thermal ensemble for the          *
*    two-dimensional ising lattice with free boundary conditions             *
*    in a magnetic field h.                                                  *
*								             *
* Storage: The state of the lattice is stored as spins +-1 in elements       *
*    [1..N][1..N] of a 2-d array of size (N+2) by (N+2).  The free boundary  *
*    conditions are handled by fixing elements [0][j], [N+1][j], [i][0],     *
*    and [i][N+1] of the array to be zero.                                   *
*                                                                            *
* Compile:  cc -o ising2d ising2d.c -lm -lcurses                             *
*  								             *
******************************************************************************
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <unistd.h>      /* for sleep system call */
#include <time.h>

/*
** PARAMETERS
** ----------
**   NX, NY:        The lattice is NX by NY sites.
**   ntherm:        Number of "thermalization" sweeps to do before starting.
**   VisualDisplay: 1 or 0 to turn on/off display of spins.
**   SleepTime:     seconds to pause between displays of each sweep.
*/

#define NX 64
#define NY 64

int ntherm = 200;
const int VisualDisplay = 1;
const int SleepTime = 100000;   // in microseconds

/*
**  THE LATTICE
*/

int spin[NX+2][NY+2];       // put fake spins around edge of lattice 


/* update_spin (s, env)
** -----------
**   Do a metropolis update on a spin *s whose "environment" in env.
** The environment should be set beforehand so that the dependence
** of the total beta*energy on the value of the selected spin is
** given by
**               (selected spin)*env .
**
** sweep (spin, N, beta, h)
** -----
**   Sweep once through all the sites of the lattice spin of length N,
** trying an update at each site with inverse temperature beta and external
** magnetic field parameter h.
*/

void update_spin(int *s, double env)
{
  int spin, newspin;
  double DeltaBetaE;

  spin = *s;
  newspin = ( drand48() < 0.5 ? 1 : -1 );
  DeltaBetaE = -(newspin-spin)*env;       /* beta*(E(new) - E(old)) */
  if ( DeltaBetaE <= 0 || drand48() < exp(-DeltaBetaE) ) *s = newspin; 
}

void sweep(double beta, double h)
{
  int nx,ny;
  double environment;

  for(nx=1; nx<=NX; nx++) for(ny=1; ny<=NY; ny++) {
    environment = 
      beta*(spin[nx][ny-1] + spin[nx][ny+1] + spin[nx-1][ny] + spin[nx+1][ny])
      + h;
    update_spin(&(spin[nx][ny]), environment);
  }
}

/* InitializeHot (spin)
** ====================
**   Initialize all the NX by NY spins of the lattice randomly.  Also
** initialize the two fake "boundary" spins to zero to implement free
** boundary conditions.
*/

void InitializeHot() {
  printf("Initializing system\n");
  int nx, ny; 
  for(nx=0; nx<=NX+1; nx++) for(ny=0; ny<=NY+1; ny++) {
    if (nx==0 || nx==NX+1 || ny==0 || ny==NY+1) spin[nx][ny]=0;
    else spin[nx][ny] = (drand48() < 0.5 ? 1 : -1);
  }
}

/* Magnetization
** =============
**   Return the volume average of the spin for the current spin
** configuration.
*/

double Magnetization() {
  int nx,ny,nmag;
  nmag = 0;
  for(nx=1; nx<=NX; nx++) for(ny=1; ny<=NY; ny++) nmag += spin[nx][ny];
  return( (double)nmag/(NX*NY));
}


/* DisplayLattice
** ==============
**   Print out the 2D lattice on the screen.
**   If SleepTime is > 0, then the screen is cleared before each display,
**   and the program pauses for SleepTime useconds after each display.
*/

void DisplayLattice(int sweep_number) {
  int nx, ny;

  if (SleepTime>0)  puts( "\033[2J" );
  for (nx=1; nx<=NX; nx++) {
    for (ny=1; ny<=NY; ny++) {
      if (spin[nx][ny]== 1) printf("X");
      if (spin[nx][ny]==-1) printf("-");
    }
    printf("\n");
  }
  if (sweep_number>=0)
    printf("sweep %i:   magnetization <sigma> = %lf\n", 
	   sweep_number, Magnetization());
  if (SleepTime>0) usleep(SleepTime);
  else printf("\n");
}


/* main
** ====
*/

int main()
{
  int n, nsweep, nx, ny;
  long ntotal, nmag;
  double beta, h;

  printf("Program generates a thermal ensemble a 2D Ising model of");
  printf("%ix%i spins with free boundary conditions.\n\n",
	 NX,NY);

  srand48((long)time(0));               // seed the number generator

  
  printf("Enter total number of configurations generated:\n");
  scanf("%d", &nsweep); 

  printf("Enter value of magnetic field parameter h:\n");
  scanf("%lf", &h);

  printf("Enter temperature parameter beta (= 1/kT):\n");
  scanf("%lf", &beta);

  InitializeHot();

  /* sweep ntherm times to thermalize system */
  printf("Thermalizing system, %d sweeps\n",ntherm);
  for(n=0; n<ntherm; n++) {
    sweep(beta,h);
    if (VisualDisplay &&  n%(ntherm/10)==0 ) {
      DisplayLattice(-1);
      printf("Thermalization sweep %d\n",n);
      usleep(1000000);
    }
  }

  /* now sweep through lattice nsweep times */
  nmag=ntotal=0;
  for(n=0; n<nsweep; n++) {
    if (VisualDisplay) DisplayLattice(n);
    sweep(beta,h);
    for(nx=1; nx<=NX; nx++) for(ny=1; ny<=NY; ny++) {
      nmag += spin[nx][ny];
      ntotal++;
    }
  }

  if (VisualDisplay) DisplayLattice(nsweep);
  printf("Average Magnetization: <s> = %lf\n", (double)nmag/ntotal);

  return(0);
}
